tests/unit/s2n_async_pkey_test.c

```diff
@@ -142,6 +152,122 @@ int async_pkey_store_callback(struct s2n_connection *conn, struct s2n_async_pkey
+
+    EXPECT_BYTEARRAY_EQUAL(input.data, expected_digest.data, expected_digest.size);
+    EXPECT_SUCCESS(s2n_free(&input));
+    EXPECT_SUCCESS(s2n_free(&expected_digest));
+
+    EXPECT_SUCCESS(s2n_async_pkey_op_set_output(op, test_signature_data, test_signature_size));
+
+    return S2N_SUCCESS;
+}
+
+int async_pkey_decrypt_callback(struct s2n_connection *conn, struct s2n_async_pkey_op *op)
+{
+    pkey_op = op;
+
+    s2n_async_pkey_op_type type = { 0 };
+    EXPECT_SUCCESS(s2n_async_pkey_op_get_op_type(op, &type));
+    EXPECT_EQUAL(type, S2N_ASYNC_DECRYPT);
+
+    uint32_t input_size = 0;
+    EXPECT_SUCCESS(s2n_async_pkey_op_get_input_size(op, &input_size));
+    EXPECT_EQUAL(input_size, test_encrypted_size);
+
+    struct s2n_blob input_buffer = { 0 };
+    EXPECT_SUCCESS(s2n_alloc(&input_buffer, input_size));
+
+    EXPECT_SUCCESS(s2n_async_pkey_op_get_input(op, input_buffer.data, input_buffer.size));
+    EXPECT_BYTEARRAY_EQUAL(input_buffer.data, test_encrypted_data, test_encrypted_size);
+    EXPECT_SUCCESS(s2n_free(&input_buffer));
+
+    EXPECT_SUCCESS(s2n_async_pkey_op_set_output(op, test_decrypted_data, test_decrypted_size));
+
+    return S2N_SUCCESS;
+}
+
+int s2n_async_sign_complete(struct s2n_connection *conn, struct s2n_blob *signature)
+{
+    EXPECT_NOT_NULL(conn);
+    EXPECT_NOT_NULL(signature);
+
+    EXPECT_EQUAL(signature->size, test_signature_size);
+    EXPECT_BYTEARRAY_EQUAL(signature->data, test_signature_data, test_signature_size);
+
+    return S2N_SUCCESS;
+}
+
+int s2n_async_decrypt_complete(struct s2n_connection *conn, bool rsa_failed, struct s2n_blob *decrypted)
+{
+    EXPECT_NOT_NULL(conn);
+    EXPECT_NOT_NULL(decrypted);
+    EXPECT_FALSE(rsa_failed);
+
+    EXPECT_EQUAL(decrypted->size, test_decrypted_size);
+    EXPECT_BYTEARRAY_EQUAL(decrypted->data, test_decrypted_data, test_decrypted_size);
+
+    return S2N_SUCCESS;
+}
+
+int async_pkey_invalid_input_callback(struct s2n_connection *conn, struct s2n_async_pkey_op *op)
+{
+    pkey_op = op;
+
+    EXPECT_FAILURE(s2n_async_pkey_op_get_op_type(op, NULL));
+    EXPECT_FAILURE(s2n_async_pkey_op_get_input_size(op, NULL));
+    EXPECT_FAILURE(s2n_async_pkey_op_get_input(op, NULL, 0));
+
+    uint32_t input_size = 0;
+    EXPECT_SUCCESS(s2n_async_pkey_op_get_input_size(op, &input_size));
+
+    uint8_t placeholder_buffer[] = { 0x0, 0x0, 0x0, 0x0 };
+
+    /* Buffer too small to contain data. */
+    EXPECT_FAILURE(s2n_async_pkey_op_get_input(op, placeholder_buffer, input_size-1));
+
+    EXPECT_FAILURE(s2n_async_pkey_op_set_output(op, NULL, test_signature_size));
+
+    return S2N_FAILURE;
+}
+
+int async_pkey_invalid_complete(struct s2n_connection *conn, struct s2n_blob *signature)
+{
+    /* This function should never be called. If this does get called that means the completion function was called even
+     * though there were issues with the callback. See async_pkey_invalid_input_callback. */
+    FAIL();
```

Comment:
You should probably use `FAIL_MSG` instead so that you can give some sort of failure message instead of relying solely on the comment here.

Reply:
Good point, I'll swap to this